Adding HTTP Support

anthonychernyak opened the pull request:
- Adds a HTTP config flag to allow connection in ethereum to connect via HTTP.
- Added any needed config flags.

ansermino commented on chains/ethereum/config.go:
```diff
@@ -22,6 +22,7 @@ type Config struct {
 	contract     common.Address
 	gasLimit     *big.Int
 	gasPrice     *big.Int
+	Http         bool // Config for type of connection
```
no need to export

ansermino commented on cmd/chainbridge/flags.go:
```diff
@@ -25,6 +25,11 @@ var (
+	HTTPConnectionFlag = cli.BoolFlag{
+		Name:  "http",
+		Usage: "Enables RPC HTTP Client",
+	}
```
As this only applies to ethereum, this should be part of ethereum `opts` (like `gasPrice`) rather than a global flag.

ansermino commented on core/chain.go:
```diff
@@ -21,5 +21,6 @@ type ChainConfig struct {
 	From         string            // address of key to use
 	KeystorePath string            // Location of key files
 	Insecure     bool              // Indicated whether the test keyring should be used
+	Http         bool              // Indicates whether to use http in the rpc connection
```
See comment above, doesn't apply to all chains

anthonychernyak commented:
@ansermino  I've updated the README and added a field to the current test. If you want me to change the functionality or add another test, let me know.

ansermino commented on chains/ethereum/config_test.go:
```diff
@@ -21,9 +21,10 @@ func TestParseChainConfig(t *testing.T) {
+			"contract":       "0x1234",
+			"gasLimit":       "10",
+			"gasPrice":       "20",
+			"connectionType": "rpc",
```
please test an expected true value

ansermino commented on README.md:
```diff
@@ -57,6 +57,7 @@ Ethereum chains support the following additional options:
 contract = ""0x12345..." // The address of the bridge contract
 gasPrice = "0x1234"      // Gas price for transactions 
 gasLimit = "0x1234"      // Gas limit for transactions
+connectionType = "rpc"  // The type of connection for the ethereum chain. The options are RPC and HTTP. If left blank, the default is RPC.
```
This isn't really what you've implemented. Anything except 'http' will be considered 'rpc' at this point. Also the options would be 'http' and 'ws', these both use rpc. Please update this and the parsing to ensure they match.

The way I initially pictured this was to have an `http` boolean option, much like in `ethereum.Connection`. Perhaps this would simplify things a little.

GregTheGreek commented on chains/ethereum/connection.go:
```diff
@@ -56,7 +56,13 @@ func NewConnection(cfg *Config, kp *secp256k1.Keypair) *Connection {
+		rpcClient, err = rpc.DialHTTP(c.cfg.endpoint)
+	} else {
+		rpcClient, err = rpc.DialWebsocket(c.ctx, c.cfg.endpoint, "/ws")
+	}
```
I envision this could be a bit more robust.  Try connecting via web-sockets first, if that has an error, attempt connecting to http. If that fails then fully error out. 

cc: @ansermino thoughts?

ansermino replied:
That doesn't really add robustness. I think it would be better to have control over which you want to be using.

ansermino replied:
@GregTheGreek

GregTheGreek replied:
yeah i guess that makes sense as well
